agent/agent-tooling/src/main/java/com/microsoft/applicationinsights/agent/internal/statsbeat/NetworkStatsbeat.java

```diff
@@ -59,116 +62,133 @@ public NetworkStatsbeat(CustomDimensions customDimensions) {
+          instrumentationKeyCounterMap.get(IntervalMetricsKey.create(ikey, null));
       return intervalMetrics == null ? 0L : intervalMetrics.requestSuccessCount.get();
     }
   }
 
   // only used by tests
-  long getRequestFailureCount(String ikey) {
+  long getRequestFailureCount(String ikey, int statusCode) {
     synchronized (lock) {
-      IntervalMetrics intervalMetrics = instrumentationKeyCounterMap.get(ikey);
+      IntervalMetrics intervalMetrics =
+          instrumentationKeyCounterMap.get(IntervalMetricsKey.create(ikey, statusCode));
       return intervalMetrics == null ? 0L : intervalMetrics.requestFailureCount.get();
     }
   }
 
   // only used by tests
   double getRequestDurationAvg(String ikey) {
     synchronized (lock) {
-      IntervalMetrics intervalMetrics = instrumentationKeyCounterMap.get(ikey);
+      IntervalMetrics intervalMetrics =
+          instrumentationKeyCounterMap.get(IntervalMetricsKey.create(ikey, null));
       return intervalMetrics == null ? 0L : intervalMetrics.getRequestDurationAvg();
     }
   }
 
   // only used by tests
-  long getRetryCount(String ikey) {
+  long getRetryCount(String ikey, int cause) {
     synchronized (lock) {
-      IntervalMetrics intervalMetrics = instrumentationKeyCounterMap.get(ikey);
+      IntervalMetrics intervalMetrics =
+          instrumentationKeyCounterMap.get(IntervalMetricsKey.create(ikey, cause));
       return intervalMetrics == null ? 0L : intervalMetrics.retryCount.get();
     }
   }
 
   // only used by tests
-  long getThrottlingCount(String ikey) {
+  long getThrottlingCount(String ikey, int cause) {
     synchronized (lock) {
-      IntervalMetrics intervalMetrics = instrumentationKeyCounterMap.get(ikey);
+      IntervalMetrics intervalMetrics =
+          instrumentationKeyCounterMap.get(IntervalMetricsKey.create(ikey, cause));
       return intervalMetrics == null ? 0L : intervalMetrics.throttlingCount.get();
     }
   }
 
   // only used by tests
-  long getExceptionCount(String ikey) {
+  long getExceptionCount(String ikey, String exceptionType) {
     synchronized (lock) {
-      IntervalMetrics intervalMetrics = instrumentationKeyCounterMap.get(ikey);
+      IntervalMetrics intervalMetrics =
+          instrumentationKeyCounterMap.get(IntervalMetricsKey.create(ikey, exceptionType));
       return intervalMetrics == null ? 0L : intervalMetrics.exceptionCount.get();
     }
   }
 
-  private void doWithIntervalMetrics(String ikey, String host, Consumer<IntervalMetrics> update) {
+  private void doWithIntervalMetrics(
+      String ikey, String host, @Nullable Object cause, Consumer<IntervalMetrics> update) {
     synchronized (lock) {
       IntervalMetrics intervalMetrics =
-          instrumentationKeyCounterMap.computeIfAbsent(ikey, k -> new IntervalMetrics());
+          instrumentationKeyCounterMap.computeIfAbsent(
+              IntervalMetricsKey.create(ikey, cause), k -> new IntervalMetrics());
       intervalMetrics.host = host;
```

Comment:
what do you think about `host` into the key also? it seems potentially more correct, and potentially more symmetric with the other properties

Reply:
yes, i think it's better that way.